dalle2_pytorch/dalle2_pytorch.py

```diff
@@ -423,54 +423,77 @@ def forward(
+        we are doing "embedding only" training.
 
-        text_embed, image_embed = rearrange_many((text_embed, image_embed), 'b d -> b 1 d')
+        In the case of embedding only training we will drop the text
+        encoding from the tokens entirely and train on the remaining
+        tokens mentioned in section 2.2 of the DALLE2 paper.
+        """
 
-        # whether text embedding is used for conditioning depends on whether text encodings are available for attention (for classifier free guidance, even though it seems from the paper it was not used in the prior ddpm, as the objective is different)
-        # but let's just do it right
-
-        if exists(mask):
-            not_all_masked_out = mask.any(dim = -1)
-            mask = torch.cat((mask, rearrange(not_all_masked_out, 'b -> b 1')), dim = 1)
-
-        if exists(mask):
-            mask = F.pad(mask, (0, 2), value = True) # extend mask for text embedding, noised image embedding, time step embedding, and learned query
+        # grab some contextual information
+        batch, device = image_embed.shape[0], image_embed.device
 
+        # gather tokens for casual transformer
+        text_embed, image_embed = rearrange_many(
+            (text_embed, image_embed), 'b d -> b 1 d')
         time_embed = self.time_embeddings(diffusion_timesteps)
         time_embed = rearrange(time_embed, 'b d -> b 1 d')
+        learned_queries = repeat(self.learned_query, 'd -> b 1 d', b=batch)
+
+        # check for embedding only training
+        if text_encodings is not None:
+            text_enc_len = text_encodings.shape[-2]
+
+            # text conditioning can be passed during training for CFG; as described in the paper
+
+            if exists(mask):
+                not_all_masked_out = mask.any(dim=-1)
+                mask = torch.cat(
+                    (mask, rearrange(not_all_masked_out, 'b -> b 1')), dim=1)
+                # extend mask for text embedding, noised image embedding, etc.
+                mask = F.pad(mask, (0, 2), value=True)
+
+            if not exists(mask):
+                mask = torch.ones((batch, text_enc_len),
+                                  device=device, dtype=torch.bool)
+
+            # classifier free guidance
+            cond_prob_mask = prob_mask_like(
+                (batch,), cond_drop_prob, device=device)
+            mask &= rearrange(cond_prob_mask, 'b -> b 1')
+
+            # actually combine everything as described originally
+            tokens = torch.cat((
+                text_encodings,
+                text_embed,
+                time_embed,
+                learned_queries
+            ), dim=-2)
+
+        # TODO: someone should double check this block of code below
+        else:
+            # during embedding only, just drop text_encodings entirely
+            tokens = torch.cat((
+                text_embed,
+                time_embed,
+                learned_queries
+            ), dim=-2)
```

Comment:
I'd like someone to check this part (above) and verify that we can simply use a mask of all ones in the shape of `(batch, tokens.shape[1])`.

My understanding is that the mask is only used for the classifier-free guidance training (text_encoding). But I could be wrong!

Reply:
yup this looks good! 💯